### PR TITLE
feat(merge-tree): add basic perf tests

### PR DESCRIPTION
**Pull request by connorskees**
## Description

Adds basic benchmarks for merge-tree insertion and removal

### CODE REVIEW

File: packages/dds/merge-tree/src/test/Removal.perf.tests.ts
Comment (Abe27342) [inline packages/dds/merge-tree/src/test/Removal.perf.tests.ts:33]: Unless things have changed since I last looked at it, these tests as well as insert tests may not be checking what you think--benchmark runs each function a number of times per cycle, and also runs a number of cycles (it uses statistical properties to determine those numbers). So here the size of the tree won't be controlled for when applying the operation.

Comment (Abe27342) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts:33]: you may want to play around with a profiler to see if that's still true. there's an `onCycle` event that you could in theory use to reset test state without it getting profiled, but I believe there are many executions of `benchmarkFn` in each cycle.

Comment (connorskees) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts]: For the insertion tests, I argue that although the tree is changing between individual iterations, the exact operation being performed on a whole is the same between separate executions, and so it should still be possible to track regressions in insertion speed. The statistics will be a bit off, but we remove the (high) overhead of having to construct the tree in every iteration, which adds more noise.

That said, I'm fine changing this to include tree construction as part of the benchmark if asked.

For the destructive tests, I'm not aware of a way around constructing the tree on every cycle.

Comment (Abe27342) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts]: benchmark.js runs each performance test for a number of cycles. For each cycle past a warmup period, it runs the test for a number of iterations and computes statistics about those samples. It then dynamically decides whether to run a new cycle and does so until one of:

1. total test time exceeds a max threshhold (default is 5s)
2. the relative margin of error (as compute by average of averages over the cycles) is below some threshhold (default is 1%)

it also keeps running cycles if there have been less than the minimum number of samples run, but in practice that isn't really an issue.

[here](https://github.com/bestiejs/benchmark.js/blob/42f3b732bac3640eddb3ae5f50e445f3141016fd/benchmark.js#L1889) is the place in benchmark source which runs on cycle completion. note it does some statistics using the data from that cycle.

This becomes an issue for insertion tests which don't do a consistent thing on each cycle, since the statistics will never really converge. If you modify your test to at least reset the tree on each cycle, the problem mostly goes away--the test is a bit misleading in that it's really measuring "the average time to insert a node for a tree of size 1000 through 1000 + iterations per cycle" (iterations per cycle seems to be 14k on my machine, but again this is dynamically chosen by benchmark). But it does at least converge and measures something closer to what you want with less error. On my machine without particularly great methodology (i was using other programs so had pretty variable cpu usage), RME is near 1% with this approach whereas without resetting the tree on cycle it stays at 5% or so.

The fact that iterations per cycle is dynamic also presents a small challenge for remove tests: if you initialize the tree to have length less than the total number of removals, you're really measuring "average time to remove all nodes of the tree of size 1000 as well as process a number of no-ops". One way you could solve this would be to construct more trees than you need at the start of each cycle (ex: construct 50 tress with length 1k if you expect iteration count to be 14k) and rotate which one you remove from. It would cause more memory pressure, but I'm curious if you'd get more consistent results.

Comment (connorskees) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts]: I've modified the insertion tests to always construct a new tree on every iteration by loading a snapshot.

Comment (connorskees) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts]: By test start, do you mean start of each iteration? Which is to say that the noise from constructing the tree inside each iteration makes the statistics not sound?

This is what the code does currently:

https://github.com/microsoft/FluidFramework/blob/2a7a12ab11f9fc31cf5bb056c9e50bd8d26c8d8b/packages/dds/merge-tree/src/test/Insertion.perf.spec.ts#L44-L49

I do like the code you've shown -- assuming there's no branch to base off of, I'll re-apply your changes to the benchmark framework in this PR and update the insertion tests to use the onCycle hook.

Comment (Abe27342) [reply packages/dds/merge-tree/src/test/Removal.perf.tests.ts]: i was referring to benchmark's notion of an iteration, which is a single run of the `benchmarkFn` callback. It runs it a number of times per cycle, and each call to `benchmark` (i.e. each test) runs a number of cycles. But yeah, exposing "onCycle" seems like an improvement. I think that code is basically good to go once you remove the `.only`.

File: packages/dds/merge-tree/src/test/PartialLengths.perf.spec.ts
Comment (Abe27342) [inline packages/dds/merge-tree/src/test/PartialLengths.perf.spec.ts:64]: nit: probably better to store the original value and restore it rather than turn it on. don't feel super strongly, as it's unlikely we'd ever change this default.

File: tools/benchmark/src/Configuration.ts
Comment (Abe27342) [inline tools/benchmark/src/Configuration.ts:195]: might be good to link to https://benchmarkjs.com/docs/#options_onCycle here

**Review by Abe27342 — APPROVED** [review]
changes to merge-tree tests look good to me. I'm not familiar with how @fluid-tools/benchmark's versioning and release is managed in this repo, so I'd double-check with someone more familiar with that setup before merging.

File: tools/benchmark/src/Configuration.ts
Comment (alexvy86) [reply tools/benchmark/src/Configuration.ts:195]: Yeah, and I'd say add our own description of what the cycle is. I found the official BenchmarkJS docs (or what I could find that wasn't StackOverflow, at least) basically non-existent, regarding the lifecycle of a test and what exactly the cycle is or what it encompasses.

**Review by alexvy86 — COMMENTED** [review]
Didn't look too in detail at the tests themselves since @Abe27342 already did, just adding a few notes.

One important thing I'd add is that the benchmark tool gets built and released independently of the client release group, so we probably want to split those changes into their own PR, merge it, publish the new version, and then do the changes in the client release group (including updating the version of benchmark used across all packages in the release group). @connorskees you tested this by npm-linking benchmark locally, I imagine? CI of this PR as it is today will never pass, because merge-tree won't find the new version of benchmark in npm.

File: packages/dds/merge-tree/package.json
Comment (alexvy86) [reply packages/dds/merge-tree/package.json]: Ah, I'd also point back to this, if we want to have the new tests run as part of the Performance Benchmarks pipeline. If they're still in flux and just for running locally, this can come later, but if we want them to run on every commit, we'll want to finish the extra steps from the linked docs.

**Comment by connorskees** [comment]
> we probably want to split those changes into their own PR

@alexvy86 The benchmark package changes have been separated in https://github.com/microsoft/FluidFramework/pull/14295. This PR hasn't yet been updated to take note of that change.

File: packages/dds/merge-tree/package.json
Comment (alexvy86) [inline packages/dds/merge-tree/package.json:82]: Benchmark is used by several packages in the client release group. pnpm does not _require_ all of the packages in the workspace to have matching versions like Lerna did, but we should probably update them in sync. I think there's a particular invocation of `flub` that would do it but don't know it by heart. Or a manual search-and-replace shouldn't be too hard, there's not that many instances.

**Pull request merged**
